runtime/src/mm/cpp/RefUpdatesTest.cpp

```diff
@@ -0,0 +1,245 @@
+testing::MockFunction<void(ObjHeader*)>* InitSingletonTest::globalConstructor_ = nullptr;
+
+} // namespace
+
+TEST_F(InitSingletonTest, InitThreadLocalSingleton) {
+    ObjHeader* location = nullptr;
+
+    ObjHeader* valueAtConstructor = nullptr;
+    EXPECT_CALL(constructor(), Call(_)).WillOnce([&location, &valueAtConstructor](ObjHeader* value) {
+        EXPECT_THAT(value, location);
+        valueAtConstructor = value;
+    });
+    ObjHeader* value = InitThreadLocalSingleton(&location, 0);
+    EXPECT_THAT(value, location);
+    EXPECT_THAT(valueAtConstructor, location);
+}
+
+TEST_F(InitSingletonTest, InitThreadLocalSingletonTwice) {
+    ObjHeader previousValue;
+    ObjHeader* location = &previousValue;
+
+    EXPECT_CALL(constructor(), Call(_)).Times(0);
+    ObjHeader* value = InitThreadLocalSingleton(&location, 0);
+    EXPECT_THAT(value, location);
+    EXPECT_THAT(value, &previousValue);
+}
+
+TEST_F(InitSingletonTest, InitThreadLocalSingletonFail) {
+    ObjHeader* location = nullptr;
+    constexpr int kException = 42;
+
+    EXPECT_CALL(constructor(), Call(_)).WillOnce([]() { throw kException; });
+    try {
+        InitThreadLocalSingleton(&location, 0);
+        ASSERT_TRUE(false); // Cannot be reached.
+    } catch (int exception) {
+        EXPECT_THAT(exception, kException);
+    }
+    EXPECT_THAT(location, nullptr);
+}
+
+TEST_F(InitSingletonTest, InitSingleton) {
+    ObjHeader* location = nullptr;
+    ObjHeader* stackLocation = nullptr;
+
+    ObjHeader* valueAtConstructor = nullptr;
+    EXPECT_CALL(constructor(), Call(_)).WillOnce([&location, &stackLocation, &valueAtConstructor](ObjHeader* value) {
+        EXPECT_THAT(value, stackLocation);
+        EXPECT_THAT(location, reinterpret_cast<ObjHeader*>(1));
+        valueAtConstructor = value;
+    });
+    ObjHeader* value = InitSingleton(&location, 0, &stackLocation);
+    EXPECT_THAT(value, stackLocation);
+    EXPECT_THAT(value, location);
+    EXPECT_THAT(valueAtConstructor, location);
+}
+
+TEST_F(InitSingletonTest, InitSingletonTwice) {
+    ObjHeader previousValue;
+    ObjHeader* location = &previousValue;
+    ObjHeader* stackLocation = nullptr;
+
+    EXPECT_CALL(constructor(), Call(_)).Times(0);
+    ObjHeader* value = InitSingleton(&location, 0, &stackLocation);
+    EXPECT_THAT(value, stackLocation);
+    EXPECT_THAT(value, location);
+    EXPECT_THAT(value, &previousValue);
+}
+
+TEST_F(InitSingletonTest, InitSingletonFail) {
+    ObjHeader* location = nullptr;
+    ObjHeader* stackLocation = nullptr;
+    constexpr int kException = 42;
+
+    EXPECT_CALL(constructor(), Call(_)).WillOnce([]() { throw kException; });
+    try {
+        InitSingleton(&location, 0, &stackLocation);
+        ASSERT_TRUE(false); // Cannot be reached.
+    } catch (int exception) {
+        EXPECT_THAT(exception, kException);
+    }
+    EXPECT_THAT(stackLocation, nullptr);
+    EXPECT_THAT(location, nullptr);
+}
+
+TEST_F(InitSingletonTest, InitSingletonRecursive) {
```

Comment:
This test looks totally cryptic to me. Why do we use two different locations for same singleton?

Reply:
It's not the same singleton - it's two different singletons. They share the constructor function due to technicalities, but the location is what identifies them. Added some comments to the test.